// ion-js/dist/amd/es6/IonLocalSymbolTable.js
define(["require", "exports", "./IonSystemSymbolTable"], function (require, exports, IonSystemSymbolTable_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    class LocalSymbolTable {
        constructor(_import = IonSystemSymbolTable_1.getSystemSymbolTableImport(), symbols = []) {
            this._import = _import;
            this.index = {};
            this._symbols = [];
            this.offset = _import.offset + _import.length;
            for (let symbol_ of symbols) {
                this.addSymbol(symbol_);
            }
        }
        get symbols() {
            return this._symbols;
        }
        get maxId() {
            return this.offset + this._symbols.length - 1;
        }
        get import() {
            return this._import;
        }
        getSymbolId(symbol_) {
            return this._import.getSymbolId(symbol_) || this.index[symbol_];
        }
        addSymbol(symbol_) {
            let existingSymbolId = this.getSymbolId(symbol_);
            if (existingSymbolId !== undefined)
                return existingSymbolId;
            let symbolId = this.offset + this.symbols.length;
            this.symbols.push(symbol_);
            this.index[symbol_] = symbolId;
            return symbolId;
        }
        getSymbolText(symbolId) {
            if (symbolId > this.maxId)
                throw new Error("SymbolID greater than maxID.");
            let importedSymbol = this.import.getSymbolText(symbolId);
            if (importedSymbol !== undefined)
                return importedSymbol;
            let index = symbolId - this.offset;
            return this.symbols[index];
        }
        numberOfSymbols() {
            return this._symbols.length;
        }
    }
    exports.LocalSymbolTable = LocalSymbolTable;
    function defaultLocalSymbolTable() {
        return new LocalSymbolTable(IonSystemSymbolTable_1.getSystemSymbolTableImport());
    }
    exports.defaultLocalSymbolTable = defaultLocalSymbolTable;
});
