// ion-js/dist/amd/es6/IonSharedSymbolTable.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    class SharedSymbolTable {
        constructor(_name, _version, _symbols) {
            this._name = _name;
            this._version = _version;
            this._symbols = _symbols;
            this._idsByText = new Map();
            this._numberOfSymbols = this._symbols.length;
            for (let m = _symbols.length - 1; m >= 0; m--) {
                this._idsByText[_symbols[m]] = m;
            }
        }
        get numberOfSymbols() {
            return this._numberOfSymbols;
        }
        get name() {
            return this._name;
        }
        get version() {
            return this._version;
        }
        getSymbolText(symbolId) {
            if (symbolId < 0) {
                throw new Error(`Index ${symbolId} is out of bounds for the SharedSymbolTable name=${this.name}, version=${this.version}`);
            }
            if (symbolId >= this.numberOfSymbols) {
                return undefined;
            }
            return this._symbols[symbolId];
        }
        getSymbolId(text) {
            let symbolId = this._idsByText[text];
            if (symbolId === undefined) {
                return null;
            }
            return symbolId;
        }
    }
    exports.SharedSymbolTable = SharedSymbolTable;
});
