// ion-js/dist/amd/es6/JsbiSupport.js
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
define(["require", "exports", "jsbi"], function (require, exports, jsbi_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    jsbi_1 = __importDefault(jsbi_1);
    class JsbiSupport {
        static isZero(value) {
            return jsbi_1.default.equal(value, JsbiSupport.ZERO);
        }
        static isNegative(value) {
            return jsbi_1.default.lessThan(value, JsbiSupport.ZERO);
        }
        static bigIntFromString(text) {
            let isNegative = false;
            let magnitudeText = text.trimLeft();
            if (text.startsWith('-')) {
                isNegative = true;
                magnitudeText = text.substring(1);
            }
            let bigInt = jsbi_1.default.BigInt(magnitudeText);
            if (isNegative) {
                bigInt = jsbi_1.default.unaryMinus(bigInt);
            }
            return bigInt;
        }
        static clampToSafeIntegerRange(value) {
            if (jsbi_1.default.greaterThan(value, this.NUMBER_MAX_SAFE_INTEGER)) {
                return Number.MAX_SAFE_INTEGER;
            }
            if (jsbi_1.default.lessThan(value, this.NUMBER_MIN_SAFE_INTEGER)) {
                return Number.MIN_SAFE_INTEGER;
            }
            return jsbi_1.default.toNumber(value);
        }
        static isSafeInteger(value) {
            return jsbi_1.default.greaterThanOrEqual(value, this.NUMBER_MIN_SAFE_INTEGER)
                && jsbi_1.default.lessThanOrEqual(value, this.NUMBER_MAX_SAFE_INTEGER);
        }
    }
    exports.JsbiSupport = JsbiSupport;
    JsbiSupport.ZERO = jsbi_1.default.BigInt(0);
    JsbiSupport.ONE = jsbi_1.default.BigInt(1);
    JsbiSupport.TWO = jsbi_1.default.BigInt(2);
    JsbiSupport.NUMBER_MAX_SAFE_INTEGER = jsbi_1.default.BigInt(Number.MAX_SAFE_INTEGER);
    JsbiSupport.NUMBER_MIN_SAFE_INTEGER = jsbi_1.default.BigInt(Number.MIN_SAFE_INTEGER);
});
