// amazon-qldb-driver-nodejs/dist/src/PooledQldbSession.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Errors_1 = require("./errors/Errors");
/**
 * Represents a pooled session object. See {@linkcode QldbSessionImpl} for more details.
 */
var PooledQldbSession = /** @class */ (function () {
    /**
     * Creates a PooledQldbSession.
     * @param session The QldbSession object that represents a session to a QLDB ledger.
     * @param returnSessionToPool A lambda that is invoked when {@linkcode PooledQldbSession.close} is called.
     */
    function PooledQldbSession(session, returnSessionToPool) {
        this._session = session;
        this._returnSessionToPool = returnSessionToPool;
        this._isClosed = false;
    }
    /**
     * Close this {@linkcode PooledQldbSession} and return the underlying {@linkcode QldbSession} to the pool.
     */
    PooledQldbSession.prototype.close = function () {
        if (this._isClosed) {
            return;
        }
        this._isClosed = true;
        this._returnSessionToPool(this._session);
    };
    /**
     * Implicitly start a transaction, execute the lambda, and commit the transaction, retrying up to the
     * retry limit if an OCC conflict or retriable exception occurs.
     *
     * @param queryLambda A lambda representing the block of code to be executed within the transaction. This cannot
     *                    have any side effects as it may be invoked multiple times, and the result cannot be trusted
     *                    until the transaction is committed.
     * @param retryIndicator An optional lambda that is invoked when the `querylambda` is about to be retried due to an
     *                       OCC conflict or retriable exception.
     * @returns Promise which fulfills with the return value of the `queryLambda` which could be a {@linkcode Result}
     *          on the result set of a statement within the lambda.
     * @throws {@linkcode SessionClosedError} when this session is closed.
     */
    PooledQldbSession.prototype.executeLambda = function (queryLambda, retryIndicator) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this._throwIfClosed();
                        return [4 /*yield*/, this._session.executeLambda(queryLambda, retryIndicator)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Implicitly start a transaction, execute the statement, and commit the transaction, retrying up to the
     * retry limit if an OCC conflict or retriable exception occurs.
     *
     * @param statement The statement to execute.
     * @param parameters An optional list of QLDB writers containing Ion values to execute.
     * @param retryIndicator An optional lambda that is invoked when the `statement` is about to be retried due to an
     *                       OCC conflict or retriable exception.
     * @returns Promise which fulfills with a Result.
     * @throws {@linkcode SessionClosedError} when this session is closed.
     */
    PooledQldbSession.prototype.executeStatement = function (statement, parameters, retryIndicator) {
        if (parameters === void 0) { parameters = []; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this._throwIfClosed();
                        return [4 /*yield*/, this._session.executeStatement(statement, parameters, retryIndicator)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Return the name of the ledger for the session.
     * @returns Returns the name of the ledger as a string.
     */
    PooledQldbSession.prototype.getLedgerName = function () {
        this._throwIfClosed();
        return this._session.getLedgerName();
    };
    /**
     * Return the session token for this session.
     * @returns Returns the session token as a string.
     */
    PooledQldbSession.prototype.getSessionToken = function () {
        this._throwIfClosed();
        return this._session.getSessionToken();
    };
    /**
     * Lists all tables in the ledger.
     * @returns Promise which fulfills with an array of table names.
     */
    PooledQldbSession.prototype.getTableNames = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this._throwIfClosed();
                        return [4 /*yield*/, this._session.getTableNames()];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Start a transaction using an available database session.
     * @returns Promise which fulfills with a Transaction object.
     * @throws {@linkcode SessionClosedError} when this session is closed.
     */
    PooledQldbSession.prototype.startTransaction = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this._throwIfClosed();
                        return [4 /*yield*/, this._session.startTransaction()];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Check and throw if this session is closed.
     * @throws {@linkcode SessionClosedError} when this session is closed.
     */
    PooledQldbSession.prototype._throwIfClosed = function () {
        if (this._isClosed) {
            throw new Errors_1.SessionClosedError();
        }
    };
    return PooledQldbSession;
}());
exports.PooledQldbSession = PooledQldbSession;
