// ion-js/dist/es6/es6/IonImport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Import {
    constructor(parent, symbolTable, length) {
        this._parent = parent;
        this._symbolTable = symbolTable;
        this._offset = this.parent ? this.parent.offset + this.parent.length : 1;
        this._length = length || this.symbolTable.numberOfSymbols;
    }
    get parent() {
        return this._parent;
    }
    get offset() {
        return this._offset;
    }
    get length() {
        return this._length;
    }
    get symbolTable() {
        return this._symbolTable;
    }
    getSymbolText(symbolId) {
        if (this.parent === undefined)
            throw new Error("Illegal parent state.");
        if (this.parent !== null) {
            let parentSymbol = this.parent.getSymbolText(symbolId);
            if (parentSymbol) {
                return parentSymbol;
            }
        }
        let index = symbolId - this.offset;
        if (index >= 0 && index < this.length) {
            return this.symbolTable.getSymbolText(index);
        }
        return undefined;
    }
    getSymbolId(symbolText) {
        let symbolId;
        if (this.parent !== null) {
            symbolId = this.parent.getSymbolId(symbolText);
            if (symbolId) {
                return symbolId;
            }
        }
        symbolId = this.symbolTable.getSymbolId(symbolText);
        if (symbolId !== null
            && symbolId !== undefined
            && symbolId < this.length) {
            return symbolId + this.offset;
        }
        return undefined;
    }
}
exports.Import = Import;
