// ion-js/dist/commonjs/es6/IonCatalog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IonSystemSymbolTable_1 = require("./IonSystemSymbolTable");
function byVersion(x, y) {
    return x.version - y.version;
}
class Catalog {
    constructor() {
        this.symbolTables = {};
        this.add(IonSystemSymbolTable_1.getSystemSymbolTable());
    }
    add(symbolTable) {
        if (symbolTable.name === undefined || symbolTable.name === null)
            throw new Error("SymbolTable name must be defined.");
        let versions = this.symbolTables[symbolTable.name];
        if (versions === undefined)
            this.symbolTables[symbolTable.name] = [];
        this.symbolTables[symbolTable.name][symbolTable.version] = symbolTable;
    }
    getVersion(name, version) {
        let tables = this.symbolTables[name];
        if (!tables)
            return null;
        let table = tables[version];
        if (!table)
            table = tables[tables.length];
        return table ? table : null;
    }
    getTable(name) {
        let versions = this.symbolTables[name], table;
        if (versions === undefined)
            return null;
        return versions[versions.length - 1];
    }
}
exports.Catalog = Catalog;
